server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

```diff
@@ -0,0 +1,473 @@
+     * {@link #DATA_STREAMS_AUTO_SHARDING_EXCLUDES_SETTING}, or if the cluster doesn't have the feature available
+     * - increase number of shards if the optimal number of shards it deems necessary for the provided data stream is GT the current number
+     * of shards
+     * - decrease the number of shards if the optimal number of shards it deems necessary for the provided data stream is LT the current
+     * number of shards
+     *
+     * The NOT_APPLICABLE type result will report a cooldown period of TimeValue.MAX_VALUE.
+     * The INCREASE/DECREASE auto sharding results will either report a cool down period that's GT TimeValue.ZERO or TimeValue.ZERO if
+     * the auto sharding can be applied now.
+     *
+     * The NO_CHANGE_REQUIRED type will always report a cool down period of TimeValue.ZERO (as there'll be no new auto sharding event)
+     */
+    public AutoShardingResult calculate(ClusterState state, DataStream dataStream, @Nullable Double writeIndexLoad) {
+        Metadata metadata = state.metadata();
+        if (isAutoShardingEnabled == false) {
+            logger.debug("Data stream auto sharding service is not enabled.");
+            return new AutoShardingResult(AutoShardingType.NOT_APPLICABLE, 0, 0, TimeValue.MAX_VALUE, null);
+        }
+
+        if (featureService.clusterHasFeature(state, DataStreamAutoShardingService.DATA_STREAM_AUTO_SHARDING_FEATURE) == false) {
+            logger.debug(
+                "Data stream auto sharding service cannot compute the optimal number of shards for data stream [{}] because the cluster "
+                    + "doesn't have the auto sharding feature",
+                dataStream.getName()
+            );
+            return new AutoShardingResult(AutoShardingType.NOT_APPLICABLE, 0, 0, TimeValue.MAX_VALUE, null);
+        }
+
+        if (dataStreamExcludePatterns.stream().anyMatch(pattern -> Regex.simpleMatch(pattern, dataStream.getName()))) {
+            logger.debug(
+                "Data stream [{}] is excluded from auto sharding via the [{}] setting",
+                dataStream.getName(),
+                DATA_STREAMS_AUTO_SHARDING_EXCLUDES_SETTING.getKey()
+            );
+            return new AutoShardingResult(AutoShardingType.NOT_APPLICABLE, 0, 0, TimeValue.MAX_VALUE, null);
+        }
+
+        if (writeIndexLoad == null) {
+            logger.debug(
+                "Data stream auto sharding service cannot compute the optimal number of shards for data stream [{}] as the write index "
+                    + "load is not available",
+                dataStream.getName()
+            );
+            return new AutoShardingResult(AutoShardingType.NOT_APPLICABLE, 0, 0, TimeValue.MAX_VALUE, null);
+        }
+        return innerCalculate(metadata, dataStream, writeIndexLoad, nowSupplier);
+    }
+
+    private AutoShardingResult innerCalculate(Metadata metadata, DataStream dataStream, double writeIndexLoad, LongSupplier nowSupplier) {
```

Comment:
It feels like this function `innerCalculate` is really two or three functions, and maybe should be split up that way (since it's fairly large). One for checking for an optimal shard increase, one for checking whether there's still a remaining decrease cooldown, and one for checking for the optimal shard decrease. What do you think?

Reply:
Good shout, done.